$patch: delete on an object results: signal SIGSEGV: segmentation violation

`$patch: delete` on an object results:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1638cf1]

goroutine 1 [running]:
sigs.k8s.io/kustomize/api/filters/fieldspec.Filter.filter(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00226d6e0, 0xd, 0x0, 0xc00052c2a0, ...)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/filters/fieldspec/fieldspec.go:70 +0x91
sigs.k8s.io/kustomize/api/filters/fieldspec.Filter.Filter(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00226d6e0, 0xd, 0x0, 0xc00052c2a0, ...)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/filters/fieldspec/fieldspec.go:52 +0x188
[...]
```

Example
workloads.yaml:
```
---
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  labels:
    role: alert-rules
  name: rule1
spec:
  groups:
  - name: rabbitmq.rules

---
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  labels:
    role: alert-rules
  name: rule2
spec:
  groups:
  - name: rabbitmq.rules
```

kustomization.yaml:
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- workloads.yaml
patches:
- patch: |
    apiVersion: monitoring.coreos.com/v1
    kind: PrometheusRule
    metadata:
      name: rule1
    $patch: delete
```

Execution:
```
$ kustomize build
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1638cf1]

goroutine 1 [running]:
sigs.k8s.io/kustomize/api/filters/fieldspec.Filter.filter(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00226d6e0, 0xd, 0x0, 0xc00052c2a0, ...)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/filters/fieldspec/fieldspec.go:70 +0x91
sigs.k8s.io/kustomize/api/filters/fieldspec.Filter.Filter(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00226d6e0, 0xd, 0x0, 0xc00052c2a0, ...)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/filters/fieldspec/fieldspec.go:52 +0x188
sigs.k8s.io/kustomize/kyaml/yaml.(*RNode).Pipe(0xc0005408c0, 0xc000286df0, 0x1, 0x1, 0xc000263b80, 0xc000263b80, 0x17fde40)
	/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.9/yaml/rnode.go:298 +0x86
sigs.k8s.io/kustomize/kyaml/yaml.(*RNode).PipeE(0xc0005408c0, 0xc000286df0, 0x1, 0x1, 0x0, 0xc000286e80)
	/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.9/yaml/rnode.go:309 +0x4d
sigs.k8s.io/kustomize/api/filters/prefixsuffix.Filter.run(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/filters/prefixsuffix/prefixsuffix.go:31 +0x186
sigs.k8s.io/kustomize/kyaml/yaml.FilterFunc.Filter(0xc00052c230, 0xc0005408c0, 0xc000054480, 0x0, 0xc00052c230)
	/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.9/yaml/types.go:118 +0x30
sigs.k8s.io/kustomize/kyaml/kio.FilterAll.func1(0xc00000e428, 0x1, 0x1, 0xc000949820, 0xc000287050, 0x149d34d, 0x179c1a0, 0xc000949820)
	/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.9/kio/kio.go:142 +0x75
sigs.k8s.io/kustomize/kyaml/kio.FilterFunc.Filter(0xc000949820, 0xc00000e428, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0xc0002870e8)
	/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.9/kio/kio.go:60 +0x44
sigs.k8s.io/kustomize/api/filters/prefixsuffix.Filter.Filter(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/filters/prefixsuffix/prefixsuffix.go:27 +0xc3
sigs.k8s.io/kustomize/api/resource.(*Resource).ApplyFilter(0xc0000945a0, 0x19307a0, 0xc00052c1c0, 0xc00052c1c0, 0x0)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/resource/resource.go:488 +0xb2
sigs.k8s.io/kustomize/api/builtins.(*PrefixSuffixTransformerPlugin).Transform(0xc0003c0680, 0x1952820, 0xc0003ee540, 0x0, 0x0)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/builtins/PrefixSuffixTransformer.go:76 +0x3b7
sigs.k8s.io/kustomize/api/internal/target.(*multiTransformer).transform(0xc000949460, 0x1952820, 0xc0003ee540, 0xc00034dd01, 0x5)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/internal/target/multitransformer.go:40 +0x79
sigs.k8s.io/kustomize/api/internal/target.(*multiTransformer).Transform(0xc000949460, 0x1952820, 0xc0003ee540, 0xc00034ddb0, 0x5)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/internal/target/multitransformer.go:35 +0x85
sigs.k8s.io/kustomize/api/internal/accumulator.(*ResAccumulator).Transform(...)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/internal/accumulator/resaccumulator.go:141
sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).runTransformers(0xc000172f40, 0xc0003ee520, 0x0, 0x0)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/internal/target/kusttarget.go:270 +0x24a
sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).accumulateTarget(0xc000172f40, 0xc0003ee520, 0xc000475700, 0x1481e27, 0x0)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/internal/target/kusttarget.go:195 +0x28d
sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).AccumulateTarget(0xc000172f40, 0x0, 0xffffffffffffffff, 0x0)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/internal/target/kusttarget.go:156 +0xce
sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).makeCustomizedResMap(0xc000172f40, 0xc00048ae01, 0x0, 0x0, 0x1f)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/internal/target/kusttarget.go:111 +0x2f
sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).MakeCustomizedResMap(...)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/internal/target/kusttarget.go:107
sigs.k8s.io/kustomize/api/krusty.(*Kustomizer).Run(0xc000287d20, 0x1837555, 0x1, 0x0, 0x0, 0x0, 0x0)
	/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.8.0/krusty/kustomizer.go:82 +0x31b
sigs.k8s.io/kustomize/kustomize/v3/commands/build.(*Options).RunBuild(0xc00015e620, 0x192eec0, 0xc0000aa008, 0x0, 0x0)
	/workspace/myClone/kustomize/commands/build/build.go:148 +0xb5
sigs.k8s.io/kustomize/kustomize/v3/commands/build.NewCmdBuild.func1(0xc0000f6dc0, 0x1e89600, 0x0, 0x0, 0x0, 0x0)
	/workspace/myClone/kustomize/commands/build/build.go:70 +0xad
github.com/spf13/cobra.(*Command).execute(0xc0000f6dc0, 0x1e89600, 0x0, 0x0, 0xc0000f6dc0, 0x1e89600)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0xc0000f6000, 0x0, 0x1739000, 0xc000096058)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
	/workspace/myClone/kustomize/main.go:14 +0x2a
```

Tested kustomize versions:
- v3.10.0
- v4.0.1

Result is similar.

--- a/api/filters/fieldspec/fieldspec.go
+++ b/api/filters/fieldspec/fieldspec.go
@@ -64,7 +64,7 @@ func (fltr Filter) filter(obj *yaml.RNode) error {
 		// found the field -- set its value
 		return fltr.SetValue(obj)
 	}
-	if obj.IsTaggedNull() {
+	if obj.IsTaggedNull() || obj.IsNil() {
 		return nil
 	}
 	switch obj.YNode().Kind {

--- a/api/internal/target/multitransformer.go
+++ b/api/internal/target/multitransformer.go
@@ -42,6 +42,14 @@ func (o *multiTransformer) transform(m resmap.ResMap) error {
 			return err
 		}
 	}
+	for _, r := range m.Resources() {
+		if r.IsEmpty() {
+			err := m.Remove(r.CurId())
+			if err != nil {
+				return err
+			}
+		}
+	}
 	return nil
 }
 

--- a/api/krusty/multiplepatch_test.go
+++ b/api/krusty/multiplepatch_test.go
@@ -1213,3 +1213,52 @@ spec:
           name: server
 `)
 }
+
+// test for #3616
+func TestSmpDeleteOnResource(t *testing.T) {
+	th := kusttest_test.MakeHarness(t)
+	th.WriteK(".", `
+resources:
+- workloads.yaml
+patches:
+- patch: |
+    apiVersion: monitoring.coreos.com/v1
+    kind: PrometheusRule
+    metadata:
+      name: rule1
+    $patch: delete
+`)
+	th.WriteF("workloads.yaml", `
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  labels:
+    role: alert-rules
+  name: rule1
+spec:
+  groups:
+  - name: rabbitmq.rules
+---
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  labels:
+    role: alert-rules
+  name: rule2
+spec:
+  groups:
+  - name: rabbitmq.rules
+`)
+	m := th.Run(".", th.MakeDefaultOptions())
+	th.AssertActualEqualsExpected(m, `
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  labels:
+    role: alert-rules
+  name: rule2
+spec:
+  groups:
+  - name: rabbitmq.rules
+`)
+}

